<a href="https://colab.research.google.com/github/macrotentional/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module1/Martin_Campbell_of_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [1]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-04-13 19:00:37--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-04-13 19:00:37 (286 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [6]:
import pandas as pd

import numpy as np

column_headers = ['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers)

print(df.shape)

df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [7]:
df = df.replace({'y': 1, 'n': 0, '?': np.NaN})
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
dem =  df[df['party']=='democrat']
rep = df[df['party']=='republican']

In [11]:
dem['aid-to-contras'].isnull().sum()

4

In [12]:
rep['aid-to-contras'].isnull().sum()

11

In [13]:
dem['aid-to-contras'].mean()

0.8288973384030418

In [14]:
rep['aid-to-contras'].mean()

0.15286624203821655

In [0]:
from scipy import stats

In [17]:
stats.ttest_ind(dem['aid-to-contras'], rep['aid-to-contras'], nan_policy='omit')

Ttest_indResult(statistic=18.052093200819733, pvalue=2.82471841372357e-54)

## Two sample t-test for the Aid to Contras bill:

1) Null hypothesis: Democrats and Republicans supported the Aid to Contras bill at equivalent rates.

2) Alternative hypothesis: Democrats and Republicans supported the Aid to Contras bill at different rates.

3) Confidence interval: 99% (.01 p-value threshold)

4) p-value: **.000000000000000000000000000000000000000000000000000000282**

5) Conclusion: based on a t-statistic of 18.052 and a p-value of  .000000000000000000000000000000000000000000000000000000282, **we can confidently reject the null hypothesis** that Democrats and Republicans supported the Aid to Contras bill at similar rates.

In [24]:
dem['right-to-sue'].mean()

0.2896825396825397

In [23]:
rep['right-to-sue'].mean()

0.8607594936708861

In [25]:
stats.ttest_ind(dem['right-to-sue'], rep['right-to-sue'], nan_policy='omit')

Ttest_indResult(statistic=-13.51064251060933, pvalue=1.2278581709672758e-34)

#Two sample t-test for the Right to Sue bill:

1) Null hypothesis: Democrats and Republicans supported the Right to Sue bill at equivalent rates.

2) Alternative hypothesis: Democrats and Republicans supported the Right to Sue bill at different rates.

3) Confidence interval: 99% (.01 p-value threshold)

4) **p-value: .000000000000000000000000000000000123**

5) Conclusion: based on a t-statistic of 13.511 and a p-value of .000000000000000000000000000000000123, **we can confidently reject the null hypothesis** that Democrats and Republicans supported the Right to Sue bill at similar rates.

In [36]:
dem['water-project'].mean()

0.502092050209205

In [37]:
rep['water-project'].mean()

0.5067567567567568

In [38]:
stats.ttest_ind(dem['water-project'], rep['water-project'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

Two sample t-test for the Water Project bill: 

1) Null hypothesis: Democrats and Republicans supported the Water Project bill at equivalent rates.

2) Alternative hypothesis: Democrats and Republicans supported the Water Project bill at different rates.

3) Confidence interval: 90% (.10 p-value threshold)

4) p-value: .929

5) Conclusion: based on a t-statistic of .089 and a p-value of .929, **we fail to reject the null hypothesis** that Democrats and Republicans supported the Water Project bill at similar rates.

## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!